substrate-archive-backend/src/database.rs

```diff
@@ -75,33 +77,68 @@ impl ReadOnlyDatabase {
             }
         }
     }
+}
+
+impl ReadOnlyDB for SecondaryRocksDB {
+    fn get(&self, col: ColumnId, key: &[u8]) -> Option<Vec<u8>> {
+        self.get(col, key)
+    }
 
-    pub fn iter<'a>(&'a self, col: u32) -> impl Iterator<Item = KeyValuePair> + 'a {
-        self.inner.iter(col)
+    fn iter<'a>(&'a self, col: u32) -> Box<dyn Iterator<Item = KeyValuePair> + 'a> {
+        Box::new(self.inner.iter(col))
     }
 
-    pub fn try_catch_up_with_primary(&self) -> Result<()> {
+    fn catch_up_with_primary(&self) -> Result<()> {
         if self.track_catchups {
             self.catch_counter.fetch_add(1, Ordering::Relaxed);
         }
         self.inner.try_catch_up_with_primary()?;
         Ok(())
     }
 
-    pub fn catch_up_count(&self) -> Option<usize> {
-        if !self.track_catchups {
-            log::warn!("catchup tracking is not enabled");
-            None
-        } else {
-            Some(self.catch_counter.fetch_add(0, Ordering::Relaxed))
+    fn open_database(
+        path: &str,
+        cache_size: usize,
+        db_path: PathBuf,
+    ) -> sp_blockchain::Result<SecondaryRocksDB> {
+        // need to make sure this is `Some` to open secondary instance
+        let db_path = db_path.as_path().to_str().expect("Creating db path failed");
+        let mut db_config = Config {
+            track_catchups: true,
+            config: DatabaseConfig {
+                secondary: Some(db_path.to_string()),
+                ..DatabaseConfig::with_columns(NUM_COLUMNS)
+            },
+        };
+        let state_col_budget = (cache_size as f64 * 0.9) as usize;
```

Comment:
I wonder if this makes sense for this use case. Like I have literally no idea what the best values are here.

Reply:
I have no idea. @insipx ?

Reply:
Been a while since I played with these values, but as far as I could tell, increasing the cache size or giving a ridiculously large cache size didn't really increase performance other than just use more memory. Inversely, giving no cache size breaks the rocksdb portions of archive.

So, in the end I went with the numbers as they were in substrate (I think this multiplication is copied right over). IIRC state column budget describes how much cache is given to each column, but we're not really accessing the same value multiple times in archive. The only time that happens is with the blocks actor/blocks column, where we might potentially iterate over the same block number in the `key` column looking for a block that is missing.